### PR TITLE
scriptcommunicator: init at 05.10

### PR DESCRIPTION
**Pull request by bignaux**
###### Motivation for this change

new

###### Things done

- [ ] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `build-use-sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [ ] NixOS
   - [ ] macOS
   - [x] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

### CODE REVIEW

File: pkgs/tools/misc/scriptcommunicator/default.nix [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+    icon = "$out/share/scriptcommunicator/images/main_32x32.png";
+    comment = "Scriptable data terminal"; 
+    desktopName = "scriptcommunicator";
+    genericName = "scriptcommunicator";
+    categories = "Network;";
+  };
+
+in stdenv.mkDerivation rec {
+
+  name = "${pname}-${version}";
```
Comment (JohnAZoidberg) [inline pkgs/tools/misc/scriptcommunicator/default.nix]: If you `inherit pname version;` you don't need to declare `name` yourself.

File: pkgs/tools/misc/scriptcommunicator/default.nix [hunk 2]
```diff
@@ -0,0 +1,87 @@
+{ stdenv
+, fetchurl, qmake, pkgconfig
+, libpng, zlib, unzip
+, qtbase, qtscript, qttools, qtserialport
+, makeWrapper, makeDesktopItem
+}:
+
+let
+  pname = "scriptcommunicator";
+  version = "05.08";
```
Comment (JohnAZoidberg) [inline pkgs/tools/misc/scriptcommunicator/default.nix]: 5.10 is out

File: pkgs/tools/misc/scriptcommunicator/default.nix [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+
+    # Install desktop file
+    mkdir -p $out/share/applications
+    ln -s ${desktopItem}/share/applications/* $out/share/applications/
+  '';
+
+  meta = with stdenv.lib; {
+    description = "Scriptable data terminal which supports several interfaces";
+    longDescription =
+    "In addition to the simple sending and receiving of data ScriptCommunicator
```
Comment (JohnAZoidberg) [inline pkgs/tools/misc/scriptcommunicator/default.nix]: Please do multiline strings like this:
```
longDescription = ''
  Foo
'';
```

File: pkgs/tools/misc/scriptcommunicator/default.nix [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+
+  nativeBuildInputs = [ unzip qmake ];
+  buildInputs = [ qtbase qtserialport qtscript makeWrapper ];
+
+  # we want QtDesigner available to create UI
+  propagatedBuildInputs = [ qttools ];
+
+  src = fetchurl {
+      url = "mirror://sourceforge/projects/scriptcommunicator/files/Source/ScriptCommunicator_05_08_source.zip";
+      sha256 = "f5ae8b1d1031c0d944fdbdd25d23288970e2e1818d214e8eec1b1f49c20f0f4d";
```
Comment (JohnAZoidberg) [inline pkgs/tools/misc/scriptcommunicator/default.nix]: Hash doesn't work anymore. Please update in conjunction with the version update.

**Review by JohnAZoidberg — CHANGES_REQUESTED** [review]
Are you still interested in getting this merged?
Please have a look at my comments. When it builds we can check whether it works as expected.

**Comment by bignaux** [comment]
@GrahamcOfBorg : build scriptcommunicator

File: pkgs/tools/misc/scriptcommunicator/default.nix [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+    owner = "szieke";
+    repo = "ScriptCommunicator_serial-terminal";
+    rev = "Release_05_10";
+    sha256 = "0v6lh5di5pxkqklp3iyvdqmr7w66c7xdd1r8lkk43sh7w7jnbvc9";
+  };
+
+  sourceRoot = "source/ScriptCommunicator";
+
+  buildPhase = ''
+    cd DeleteFolder/DeleteFolder
```
Comment (JohnAZoidberg) [inline pkgs/tools/misc/scriptcommunicator/default.nix]: Do you have to `cd` into the directories for `qmake` to work?
If you have to, I think you could use `qmake -r DeleteFolder.pro ScriptEditor.pro ScriptCommunicator.pro`

File: pkgs/tools/misc/scriptcommunicator/default.nix [hunk 6]
```diff
@@ -0,0 +1,93 @@
...
+
+    # Need to fix this - i don't populate with it for now.
+    #cp aardvarkI2cSpi/aardvark.so cheetahSpi/cheetah.so $out/lib/
+    cp -a ScriptEditor/apiFiles exampleScripts images templates test $out/share/scriptcommunicator
+    cp ScriptEditor/ScriptEditor $out/share/scriptcommunicator/
+    cp DeleteFolder/DeleteFolder/DeleteFolder $out/share/scriptcommunicator/
+    cp ScriptCommunicator $out/share/scriptcommunicator/
+    cp documentation/Manual_ScriptCommunicator.pdf $out/share/scriptcommunicator/
+    
+    makeWrapper $out/share/scriptcommunicator/ScriptCommunicator $out/bin/scriptcommunicator \
```
Comment (JohnAZoidberg) [inline pkgs/tools/misc/scriptcommunicator/default.nix]: Why not copy directly to `$out/bin` and wrap with `wrapProgram $out/bin/Scriptcommunicator --[args]`?

File: pkgs/tools/misc/scriptcommunicator/default.nix [hunk 7]
```diff
@@ -0,0 +1,93 @@
...
+, libpng, zlib, unzip
+, qt5, qttools, qtserialport, qtscript, qtmultimedia
+, makeWrapper, makeDesktopItem
+}:
+
+let
+  pname = "scriptcommunicator";
+  version = "05.10";
+
+  desktopItem = makeDesktopItem {
```
Comment (JohnAZoidberg) [inline pkgs/tools/misc/scriptcommunicator/default.nix]: Nice! Good job, thinking about that.
I think `desktopName` should be `ScriptCommunicator`,
`GenericName` should be something like "Serial Terminal" or "Data Terminal" (I like the former better),
you can set `name = pname;`
I think exec should be the absolute path but I'm not sure if `$out/bin/ScriptCommunicator` works here.

File: pkgs/tools/misc/scriptcommunicator/default.nix [hunk 8]
```diff
@@ -0,0 +1,93 @@
...
+    qmake ScriptCommunicator.pro
+    make
+  '';
+
+  installPhase = ''
+    mkdir -p $out/{bin,share/scriptcommunicator,lib}
+
+    # Need to fix this - i don't populate with it for now.
+    #cp aardvarkI2cSpi/aardvark.so cheetahSpi/cheetah.so $out/lib/
+    cp -a ScriptEditor/apiFiles exampleScripts images templates test $out/share/scriptcommunicator
```
Comment (JohnAZoidberg) [inline pkgs/tools/misc/scriptcommunicator/default.nix]: Do we need test?

**Comment by JohnAZoidberg** [comment]
Builds and runs for me :)
I don't have any serial hardware though - have you tried if it actually does the right thing?

File: pkgs/tools/misc/scriptcommunicator/default.nix (hunk 8 above)
Comment (bignaux) [reply pkgs/tools/misc/scriptcommunicator/default.nix]: anyway if we would test really , we need special hardware, plus i think cheetahspi is closed source and globally, this scriptcommunicator is not written by a developer using linux so things are a bit ugly on the build system. If people need it, he has to write some PR. i find this tool very interesting in the feature (qt script, qt ui and so) and i think 99% of possible user on nixos would use it for simple serial that i test. So no really harmful.

Comment (JohnAZoidberg) [reply pkgs/tools/misc/scriptcommunicator/default.nix]: Yeah, I noticed - it's rough...
Sure, I'd be happy if you can confirm that the main functionality works.
I'd change the commend to something like `# Cheetah and its wrapper are proprietary libraries and need special hardware`

File: pkgs/tools/misc/scriptcommunicator/default.nix [hunk 9]
```diff
@@ -0,0 +1,93 @@
...
+  nativeBuildInputs = [ unzip qmake makeWrapper ];
+  buildInputs = [  qtserialport qtscript qtmultimedia ];
+
+  # we want QtDesigner available to create UI
+  propagatedBuildInputs = [ qttools ]; 
+
+  src = fetchFromGitHub {
+    owner = "szieke";
+    repo = "ScriptCommunicator_serial-terminal";
+    rev = "Release_05_10";
```
Comment (jtojnar) [inline pkgs/tools/misc/scriptcommunicator/default.nix]:
```suggestion
    rev = "Release_${stdenv.lib.replaceStrings ["."] ["_"] version}";
```

**Review by jtojnar — CHANGES_REQUESTED** [review]

File: pkgs/tools/misc/scriptcommunicator/default.nix (hunk 5 above)
Comment (bignaux) [reply pkgs/tools/misc/scriptcommunicator/default.nix]: No way, i don't success to do that.

File: pkgs/tools/misc/scriptcommunicator/default.nix (hunk 8 above)
Comment (bignaux) [reply pkgs/tools/misc/scriptcommunicator/default.nix]: fixed according to author.

**Comment by bignaux** [comment]
Please merge it like it is for the moment, i'll improve later with help of the author.

File: pkgs/tools/misc/scriptcommunicator/default.nix (hunk 5 above)
Comment (aanderse) [reply pkgs/tools/misc/scriptcommunicator/default.nix]: `pushd` and `popd` might be a bit nicer to use here

Comment (aanderse) [reply pkgs/tools/misc/scriptcommunicator/default.nix]: @bignaux ping

Comment (bignaux) [reply pkgs/tools/misc/scriptcommunicator/default.nix]: i don't have time to make a perfect recipe for this bugging tool. I improve derivation on new release, time to time. Better is the enemy of good.

Comment (aanderse) [reply pkgs/tools/misc/scriptcommunicator/default.nix]: @bignaux no worries. Just trying to be helpful and promote consistent coding style within nixpkgs.

**Pull request closed without merging**